[#84] Write intro tutorial (#115)
kowainik/co-log: co-log/example/1-intro/Intro.md

# Intro: Using `LogAction`

This tutorial is an introduction to `co-log`. It contains basic examples of
using core data types and functions.

You can run this tutorial by calling the following command:

```shell
cabal new-run intro
```

## Preamble: imports and language extensions

Since this is a literate haskell file, we need to specify all our language
extensions and imports up front.

```haskell
import Colog.Core (LogAction (..), (<&), logStringStdout)
```

## Core data type

`co-log` is based on the following data type:

```idris
newtype LogAction m msg = LogAction
    { unLogAction :: msg -> m ()
    }
```

Logging action is a function from some message of user-defined type `msg` that
performs all logic inside some monad `m`. In the `co-log` library **logger** is represented as a
value. With such approach, you can modify the way you do logging by simply performing some
transformations with the value you have.

> Let's first look at a very basic example of using just `putStrLn` for logging:

Reviewer: ```suggestion
Let's first look at a very basic example of simply using `putStrLn` for logging:
```

```haskell
example0 :: IO ()
example0 = do
    putStrLn "Example 0: First message"
    putStrLn "Example 0: Second message"
```

Using `putStrLn` for logging is a very simple and basic approach for logging.
When your application becomes bigger and more complex, you might want to bring

> some logging library into it.

Reviewer: probably add reasons why would you want that, like, to let the library handle more difficult cases or anything like that?

> Now let's look at how you can `LogAction` instead of `putStrLn` to achieve the

Reviewer: ```suggestion
Now let's look at how you can use `LogAction` instead of `putStrLn` to achieve the
```

> same goal. With `co-log` you need to have value of type `LogAction` that defines

Reviewer: ```suggestion
same goal. With `co-log` you need to have a value of type `LogAction` that defines
```

> how you are going to do logging. So you configure your logging separately and

Reviewer: ```suggestion
how you are going to do logging. So you configure your logging settings separately and
```

then pass and use this `LogAction` value. See the following example for more

> details:

Reviewer: Probably, it's better to point how this function is almost the same as the example0 and say something about how `logger` could look like, or say that you will cover it later? I guess it could help understand all together better, or how do you think?

Author: I think it's a good idea to say how the logging can look like, so readers can have better understanding. Thanks for the suggestion!

```haskell
example1 :: LogAction IO String -> IO ()
example1 logger = do
    unLogAction logger "Example 1: First message"
    unLogAction logger "Example 1: Second message"
```

> If you want to do logging with `co-log`, then of the options (and the simplest one)

Reviewer: ```suggestion
If you want to do logging with `co-log`, then one of the options (and the simplest one)
```

# Intro: Using `LogAction`

This tutorial is an introduction to `co-log`. It contains basic examples of
using core data types and functions.

You can run this tutorial by calling the following command:

```shell
cabal new-run intro
```

## Preamble: imports and language extensions

Since this is a literate haskell file, we need to specify all our language
extensions and imports up front.

```haskell
import Colog.Core (LogAction (..), (<&), logStringStdout)
```

## Core data type

`co-log` is based on the following data type:

```idris
newtype LogAction m msg = LogAction
    { unLogAction :: msg -> m ()
    }
```

Logging action is a function from some message of user-defined type `msg` that

> perform all logic inside some monad `m`. In `co-log` logging represented as a

Reviewer: ```suggestion
perform all logic inside some monad `m`. In the `co-log` library logging is represented as a
```

> value. With approach you can modify the way you do logging by simply performing

Reviewer: ```suggestion
value. With such approach, you can modify the way you do logging by simply performing some
```

transformations with the value you have.

> Let's first look at very basic example of using `LogAction`:

Reviewer: ```suggestion
Let's first look at a very basic example of using `LogAction`:
```

```haskell
example1 :: LogAction IO String -> IO ()
example1 logger = do
    unLogAction logger "Example 1: First message"
    unLogAction logger "Example 1: Second message"
```

> If you want to do logging, you need to pass `LogAction` as an argument to your

Reviewer: This is a too strict statement. Probably better to say that passing the arguments id only **one of** the options, and it is the simplest one.

> function. In this example we used `LogAction` that takes `String`s as messages

Reviewer: ```suggestion
function. In this example, we are using `LogAction` that takes `String`s as messages
```

and performs logging inside `IO` monad.

> For convenience, library defines useful operator `<&` that makes logging easier:

Reviewer: Probably not logging, but it makes code easier?

```haskell
example2 :: LogAction IO String -> IO ()
example2 logger = do
    logger <& "Example 2: First message"
    logger <& "Example 2: Second message"
```

> In order to do some logging, we need to pass some `logger` to our functions.

Reviewer: Before you already wrote that you do logging, but here you say it like this is the function that would make logging, not the previous ones, probably could be rephrased, or some previous sentences should be modified? I'm a bit lost what is logging and what is not.

Here we are going to use the following `LogAction`:

```idris
logStringStdout :: LogAction IO String
```

> This action just uses `putStrLn` underhood.

Reviewer: ```suggestion
This action uses `putStrLn` underhood.
```

Reviewer: Also probably mention `stdout`.

```haskell
main :: IO ()
main = do
    let logger = logStringStdout
    example1 logger
    example2 logger

> ```

Reviewer: Probably comments before this piece of code about what is going on (like putting all together)?
